### ISSUE TITLE
Enable gather vectorization in LLVMGPU backend

### ISSUE THREAD

**Issue by ThomasRaoux** [iree-org/iree#12737]
### Request description

linalg operations with tensor.extract can now be vectorized, this means we can vectorize gather operations.

The following gather operation is common in Bert/gpt kind of model:

```
func.func @foo(%arg1: tensor<50257x768xf32>, %12: tensor<1x5x1xi32>) -> tensor<1x5x768xf32> {
  %13 = "mhlo.gather"(%arg1, %12) 
  {dimension_numbers = #mhlo.gather<offset_dims = [2],
  collapsed_slice_dims = [0],
  start_index_map = [0],
  index_vector_dim = 2>,
  indices_are_sorted = false,
  slice_sizes = dense<[1, 768]> : tensor<2xi64>} : 
  (tensor<50257x768xf32>, tensor<1x5x1xi32>) -> tensor<1x5x768xf32>
  return %13 : tensor<1x5x768xf32>
}
```

It can be compiled with the following command line:
`../iree-build/tools/iree-compile --iree-hal-target-backends=cuda gather.mlir -o gather.vmfb  -iree-input-type=mhlo`

IREE currently doesn't vectorize it because we are not enabling vectorization of linalg op with extract. This is something we can turn on.

We most likely need to refactor the way we call vectorization here:
https://github.com/openxla/iree/blob/d9b052d57713122b8c74a89aa4f383b3063105d4/compiler/src/iree/compiler/Codegen/Common/GPUVectorization.cpp#L59

It could be as simple as propagating the option to vectorize ND extract.

We want this option to be turned on:
https://github.com/llvm/llvm-project/blob/c39dd7c1db97fa367cb6282067b74cd8e55ef09a/mlir/include/mlir/Dialect/Linalg/Transforms/Transforms.h#L567

### What component(s) does this issue relate to?

Compiler

**Comment by pavanimajety** [comment]
The pipeline currently executes -- `WorkgroupSpecialization` --> `GPUTileTensor`  --> `GPUVectorization` transforms in order of execution. `GPUTileTensor` is introducing dynamic dimensions which prevent further vectorization of `linalg.generic` op.  
Here is a gist of the IR. https://gist.github.com/pavanimajety/7e608924a2613dadaae7b1f7a38705a3

However, if I flip the order of execution, I am able to vectorize the `linalg.generic` op but tiling is not working. 
https://gist.github.com/pavanimajety/1a1b9c72a7b17a03f775fb49bf8912b0

@ThomasRaoux What is expected in this case? Do we have a preference of which transform needs to run before the other transform or do we have any option to disable tiling when vectorization is more preferred transform of the two?

**Comment by pjannaty** [comment]
Also cc @MaheshRavishankar if Mahesh has insights.

**Comment by ThomasRaoux** [comment]
> The pipeline currently executes -- `WorkgroupSpecialization` --> `GPUTileTensor` --> `GPUVectorization` transforms in order of execution. `GPUTileTensor` is introducing dynamic dimensions which prevent further vectorization of `linalg.generic` op. Here is a gist of the IR. https://gist.github.com/pavanimajety/7e608924a2613dadaae7b1f7a38705a3
> 
> However, if I flip the order of execution, I am able to vectorize the `linalg.generic` op but tiling is not working. https://gist.github.com/pavanimajety/1a1b9c72a7b17a03f775fb49bf8912b0
> 
> @ThomasRaoux What is expected in this case? Do we have a preference of which transform needs to run before the other transform or do we have any option to disable tiling when vectorization is more preferred transform of the two?

sorry for the late response. The examples is probably not ideal as the size is smaller than 64 elements which is what we use by default as number of threads per block. Using a larger size should should the problem. It sounds like you were able to try it on different examples?

**Comment by pavanimajety** [comment]
Closed via #13358
